pkg/grpc_middleware/dialer/builder.go

```diff
@@ -0,0 +1,319 @@
+// flexibility if a consumer wants to set a default and or override
+func (b *Builder) SetConnInfo(dns, port string, tls bool) error {
+	u, err := url.Parse(dns)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("invalid dns: %s", dns))
+	}
+	if u.Scheme != "" {
+		// This will cause the connection to fail silently with a cryptic "context.Deadline exceeded" so we validate for it here.
+		return errors.New(fmt.Sprintf("grpc connection dns must not contain a scheme/protocol: %s", dns))
+	}
+	b.dns = &dns
+
+	i, err := strconv.ParseUint(port, 10, 32)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("invalid port number: %s", port))
+	}
+	if i > 65535 {
+		return errors.Errorf("invalid port number: %s", port)
+	}
+	b.port = &port
+	b.withTLS = &tls
+
+	return nil
+}
+
+// GetConnInfo returns the dns and port that were set, and errors if either
+// were null, allowing verfication prior to attempting the connection
+func (b *Builder) GetConnInfo() (dns string, port string, useTLS bool, err error) {
+	if b.dns == nil || b.port == nil || b.withTLS == nil {
+		return "", "", false, errors.New("Connection info not fully set")
+	}
+	return *b.dns, *b.port, *b.withTLS, nil
+}
+
+// SetConnAddr sets connection information as well as configures some common tls and authentication options
+// based on a provided connection string.
+// the format is tls://hostname:port (or tcp://hostname:port if developing locally without TLS)
+// Optional TLS parameters:
+//   skip_verify=true        ignore server CA verification.
+//   ca_file=./filename.pem  CA of service for verification.
+func (b *Builder) SetConnAddr(addr string) error {
+	cfg, err := ReadConfig(addr)
+	if err != nil {
+		return err
+	}
+	cfg.ApplyToBuilder(b)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Dial returns the client connection to the server.
+// context is ignored unless builder is set to block using WithBlock(true)
+func (b *Builder) Dial(ctx context.Context) (*grpc.ClientConn, error) {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	dns, port, withTLS, err := b.GetConnInfo()
+	if err != nil {
+		return nil, fmt.Errorf("target connection parameter missing: dns and/or port not set")
+	}
+
+	addr := dns + ":" + port
+	log.Printf("Target to connect = %s, tls = %t", addr, withTLS)
+
+	options := b.joinOptions(withTLS)
+
+	cc, err := grpc.DialContext(ctx, addr, options...)
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to client. error = %+v", err)
+	}
+	return cc, nil
+}
+
+// DialAddr returns the client connection to the server
+// context is ignored unless builder is set to block using WithBlock(true)
+func (b *Builder) DialAddr(ctx context.Context, addr string) (*grpc.ClientConn, error) {
+	if err := b.SetConnAddr(addr); err != nil {
+		return nil, err
+	}
+
+	return b.Dial(ctx)
+}
+
+// Clone the builder and return copy
+func (b *Builder) Clone() *Builder {
+	c := &Builder{
```

Comment:
what's the purpose of this?

Reply:
This is used to clone a copy of the builder. An example of its use would be in something like the graphql-gateway passing a builder to multiple microservices which may have their own changes to the client. Those changes do not effect each other. 

see my example here https://github.com/caring/graphql-gateway/blob/jlundy/new-grpc-dialer/cmd/server/config.go#L87-L92

i started with a base Builder that sets the user agent for all clients generated from the builder. But then for client specific settings i need to clone so they don't trample each other.

Reply:
If I understand what you're saying... graphQL gateway would spin up the builder once, assuming it's largely consistent between services....

Then for each microservice, clone it to change just the connection info?

Reply:
sort of yes.. if every gateway microservice had to add its own snowflakey bits to their connection. I am thinking intercepters and keepalives maybe. Connection info is maybe a bit convoluted. if that was the only difference then one should just pass the connection. 

with passing a builder the lifetime now becomes indeterminate. you don't know when the microservice will call Dial or if the microservice will add conflicting options. Which can lead to race issues if you gave multiple microservices the same builder.

Reply:
thus why I raised the question of benefit in the first place... wasn't sure of use cases where we'd need to clone/modify/use instead of just building anew

Reply:
I don't think in your example you _need_ the clone functionality, since you're only using that builder once...

Reply:
true.. i added to my example above. :)

Reply:
I suppose a more lightweight version of the clone would be to wrap the builder in a builder.. so child builders would fall back onto the parent rather than copy. 🤔